dispatch_mem_profiler.py

```diff
@@ -0,0 +1,115 @@
+import functools
+import torch
+
+import torch
+import torch.nn as nn
+from torch.utils._pytree import tree_map, tree_flatten
+from typing import List, Any
+from numbers import Number
+from collections import defaultdict
+from torch.utils._python_dispatch import push_torch_dispatch_mode, TorchDispatchMode
+aten = torch.ops.aten
+
+MB = 1024 * 1024.0
+
+operator_names = defaultdict(int)
+mem_usage = defaultdict(float)
+markers = defaultdict(int)
+series = {}
+
+def normalize_tuple(x):
+    if not isinstance(x, tuple):
+        return (x,)
+    return x
+
+
+class MemoryProfileDispatchMode(TorchDispatchMode):
+    def __init__(self, verbose=False):
+        self.verbose = verbose
+    
+    def __torch_dispatch__(self, func, types, args=..., kwargs=None):
+
+        global mem_usage, operator_names
+        rs = func(*args, **kwargs)
+        if func == torch.ops.aten.detach.default:
+            return rs
+        mem = torch.cuda.memory_allocated()/MB
+        func_name = func.__name__ + "_"+ str(operator_names[func.__name__])
+        operator_names[func.__name__] = operator_names[func.__name__] + 1
+        mem_usage[func_name] = mem
+        if(self.verbose):
+            print("Mem Usage ("+func_name + "): ", mem)
+        return rs
+
+def clear_state():
+    global operator_names, mem_usage
+    operator_names.clear()
+    mem_usage.clear()
+
+def add_series(series_name):
+    global mem_usage, series
+    fin_usage = torch.cuda.memory_allocated()/MB
+    mem_usage['fin_usage'] = fin_usage
+    series[series_name]=mem_usage
+    mem_usage = defaultdict(float)
+
+def save_graph(filename:str):
+    global series, markers
+    import matplotlib.pyplot as plt
+    for series_name, mem_usage in series.items():
+        y = mem_usage.values()
+        min_val = min(y)
+        max_val = max(y)
+        x = [i for i in range(len(y))]
+        plt.plot(x,y, label=series_name)
+    plt.xlabel("# Operator Calls")
+    plt.ylabel("Allocated Memory (MB)")
+    plt.title(filename)
+    for marker_name,marker in markers.items():
+        plt.plot([marker, marker], [min_val, max_val], 'k-', lw=2, label=marker_name)
+    plt.legend()
+    print("Saving Graph")
+    plt.savefig("../"+filename)
+
+def add_marker(marker_name):
+    global markers, mem_usage, series
+    k = len(series.keys())
+    last_val_num = len(mem_usage.values())
+    markers[marker_name+str(k)] = last_val_num
+
+def reduce_to_scalar_loss(inp):
+    return inp.sum()
+
+from functorch.compile import aot_module, nop, print_compile, min_cut_rematerialization_partition
```

Comment:
Can you move this to the top? Also try running `black` formatter just for consistent styling.